lighthouse-core/gather/computed/page-dependency-graph.js

```diff
@@ -34,52 +38,203 @@ class PageDependencyGraphArtifact extends ComputedArtifact {
   }
 
   /**
-   * @param {!TraceOfTabArtifact} traceOfTab
    * @param {!Array<!WebInspector.NetworkRequest>} networkRecords
-   * @return {!Node}
+   * @return {!NetworkNodeOutput}
    */
-  static createGraph(traceOfTab, networkRecords) {
+  static getNetworkNodes(networkRecords) {
+    const nodes = [];
     const idToNodeMap = new Map();
     const urlToNodeMap = new Map();
 
     networkRecords.forEach(record => {
       const node = new NetworkNode(record);
+      nodes.push(node);
       idToNodeMap.set(record.requestId, node);
 
       if (urlToNodeMap.has(record.url)) {
-        // If duplicate requests have been made to this URL we can't be certain which node is being
-        // referenced, so act like we don't know the URL at all.
-        urlToNodeMap.set(record.url, undefined);
+        const list = urlToNodeMap.get(record.url);
+        list.push(node);
       } else {
-        urlToNodeMap.set(record.url, node);
+        urlToNodeMap.set(record.url, [node]);
       }
     });
 
-    const rootRequest = networkRecords
-        .reduce((min, next) => min.startTime < next.startTime ? min : next);
-    const rootNode = idToNodeMap.get(rootRequest.requestId);
-    networkRecords.forEach(record => {
-      const initiators = PageDependencyGraphArtifact.getNetworkInitiators(record);
-      const node = idToNodeMap.get(record.requestId);
+    return {nodes, idToNodeMap, urlToNodeMap};
+  }
+
+  /**
+   * @param {!TraceOfTabArtifact} traceOfTab
+   * @return {!Array<!CPUNode>}
+   */
+  static getCPUNodes(traceOfTab) {
+    const nodes = [];
+    let i = 0;
+    while (i < traceOfTab.mainThreadEvents.length) {
+      const evt = traceOfTab.mainThreadEvents[i];
+
+      // Skip all trace events that aren't schedulable tasks with sizable duration
+      if (
+        evt.name !== TracingProcessor.SCHEDULABLE_TASK_TITLE ||
+        !evt.dur ||
+        evt.dur < MINIMUM_TASK_DURATION
+      ) {
+        i++;
+        continue;
+      }
+
+      // Capture all events that occurred within the task
+      const children = [];
+      i++; // Start examining events after this one
+      for (
+        const endTime = evt.ts + evt.dur;
+        i < traceOfTab.mainThreadEvents.length && traceOfTab.mainThreadEvents[i].ts < endTime;
+        i++
+      ) {
+        children.push(traceOfTab.mainThreadEvents[i]);
+      }
+
+      nodes.push(new CPUNode(evt, children));
+    }
+
+    return nodes;
+  }
+
+  /**
+   * @param {!Node} rootNode
+   * @param {!NetworkNodeOutput} networkNodeOutput
+   */
+  static linkNetworkNodes(rootNode, networkNodeOutput) {
+    networkNodeOutput.nodes.forEach(node => {
+      const initiators = PageDependencyGraphArtifact.getNetworkInitiators(node.record);
```

Comment:
so there is a case where these initiators kind of lie to you, as far as dependencies are concerned.

example:  
* a page has a stylesheet. in that stylesheet are background images and webfonts that will be required.
* the styles have not be calculated yet, but they are downloaded. (yes, chrome should probably try to recalc style/scan/preload these resources immediately after download, but it doesn't)
* an unrelated script on the page runs and forces layout. style recalc is forced because layout.
* recalc runs and the requests for images and fonts begin.
* their initiators have a stack trace including this unrelated script (and whatever other unrelated scripts are in the call stack)

unfortunately this happens pretty often.

----------

to avoid misattributing the dependency i think we'd want to exclude a stack trace if the request began during a StyleRecalc main thread event.  :)

the good news is brendan and i couldn't think of another case where the call stack shouldn't be considered dependencies.

Reply:
I'm a little lost on which initiators you're talking about here, so you're saying that image and font network records have in their initiator information a stack trace that's not relevant and so I should cross-reference that against CPU task information to see if we're in a stylerecalc? This feels like the initiator should be fixed instead though because the initiator isn't the script it's the stylesheet, right?

Reply:
>  image and font network records have in their initiator information a stack trace that's not relevant

yes correct

>  I should cross-reference that against CPU task information to see if we're in a stylerecalc

yah that's the solution that comes to mind.

> This feels like the initiator should be fixed instead though because the initiator isn't the script it's the stylesheet, right?

If we want to be really specific there are two _initiator-type_ like ideas:  1) **the URL source resource** (bad name sorry). There is a specific resource that contains that URL (like a stylesheet). Obviously a request can't be made to it without this file being present. 2) **the initiating task.** - the main thread activity that caused the browser to request this file. There is sometimes a JS callstack associated with this, but other times it's an async task like the browser just catching up to styles it hasn't computed yet.

Oftentimes 1 and 2 point to the same url. (a script requesting another script). But this forced layout case is the primary example of when they diverge.  IMO developers want both pieces of information. 

I'd tweak your statement to say the initiator should be fixed via breaking into the two components. Then we could clearly ask for the URL source resource and it'd give us a stylesheet rather than whatever JS forced a stylerecalc to happen.


starbucks.com is an example of this behavior. see the first google webfont's initiator is a line of JS that calls `offsetWidth`. 

![screen shot 2017-09-12 at 10 37 44 am](https://user-images.githubusercontent.com/39191/30340184-7dcd743c-97a6-11e7-9040-a4e536a17833.png)

Reply:
Ahhhhhh I see, gotcha. Yeah then let's dig into this with the rest of the pending initiator improvements for #3163